gnovm/pkg/gnolang/preprocess.go

```diff
@@ -803,44 +825,48 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 								Op:    n.Op,
 								Right: rn,
 							}
-							resn := Node(Call(tx, n2))
+							resn := Node(Call(tx, n2)) // this make current node to gonative{xxx}
 							resn = Preprocess(store, last, resn)
 							return resn, TRANS_CONTINUE
 							// NOTE: binary operations are always computed in
 							// gno, never with reflect.
 						} else {
-							if isShift {
-								// nothing to do, right type is (already) uint type.
-								// we don't yet know what this type should be,
-								// but another checkOrConvertType() later does.
-								// (e.g. from AssignStmt or other).
-							} else {
+							if !isShift {
 								// convert n.Left to right type.
-								checkOrConvertType(store, last, &n.Left, rt, false)
+								checkOrConvertType(store, last, &n.Left, rt, false, false)
+							}
+						}
+					} else if lcx.T == nil { // LHS is nil. XXX, should this be located in isUntyped?
+						if !isShift {
+							// convert n.Left to typed-nil type.
+							checkOrConvertType(store, last, &n.Left, rt, false, false)
+						}
+					} else { // left is typed const, right not const
+						if !isShift {
+							if isUntyped(rt) { // refer to 0_d.gno. e.g. int(1) + 1<<x
+								checkOrConvertType(store, last, &n.Right, lt, false, false)
+							} else { // left typed const, right typed non-const
+								checkOrConvertType(store, last, &n.Left, rt, false, false)
 							}
 						}
-					} else if lcx.T == nil {
-						// convert n.Left to typed-nil type.
-						checkOrConvertType(store, last, &n.Left, rt, false)
 					}
-				} else if ric {
+				} else if ric { // right is const, left is not
 					if isUntyped(rcx.T) {
 						// Left not, Right untyped const ----------------
-						if isShift {
-							if baseOf(rt) != UintType {
-								// convert n.Right to (gno) uint type.
-								checkOrConvertType(store, last, &n.Right, UintType, false)
-							} else {
-								// leave n.Left as is and baseOf(n.Right) as UintType.
-							}
-						} else {
+						if !isShift {
```

Comment:
maybe add comment noting that for shift expressions, we already perform a conversion to uint before

Reply:
resolved